### PR TITLE
[api-minor] Allows a worker to handle multiple documents.

### PR DESCRIPTION
**Pull request by yurydelendik**
Added features:
- Allows create a PDFWorker object prior getDocument use
- Every loading task has unique `docId`, that allows generate unique fonts and use several PDF documents inside single worker
- FontLoader refactored to not be global

### CODE REVIEW

**Comment by yurydelendik** [comment]
Fixes #5992
Alternative solution for #5984
Shall address #6209,  #6205, #5148, #4228

File: src/display/api.js [hunk 1]
```diff
@@ -341,10 +363,25 @@ PDFJS.getDocument = function getDocument(src,
...
+  var nextDocumentId = 0;
+
   /** @constructs PDFDocumentLoadingTask */
   function PDFDocumentLoadingTask() {
     this._capability = createPromiseCapability();
     this._transport = null;
+    this._worker = null;
+
+    /**
+     * Unique document loading task id -- used in MessageHanlders.
```
Comment (existentialism) [inline src/display/api.js]: Nit: typo `MessageHandlers`

**Comment by yurydelendik** [comment]
Some unit tests added to test PDFWorker functionality.

/botio test

**Comment by Snuffleupagus** [comment]
I've looked through (and manually tested) this PR, and it all looks good to me!

@brendandahl Would you mind looking through this as well?

File: src/display/api.js [hunk 2]
```diff
@@ -331,27 +338,95 @@ PDFJS.getDocument = function getDocument(src,
...
 /**
+ * Starts fetching of specified PDF document/data.
+ * @param {Object} source
+ * @param {PDFDataRangeTransport} pdfDataRangeTransport
+ * @param {string} docId Unique document id, used as MessageHandler id.
+ * @returns {Promise} The promise, which is resolved when worker id of
+ *                    MessageHandler is known.
+ * @private
+ */
+function _fetchDocument(worker, source, pdfDataRangeTransport, docId) {
```
Comment (Snuffleupagus) [inline src/display/api.js]: Nit: the `worker` parameter is missing in JSDoc comments above.

File: src/display/api.js [hunk 3]
```diff
@@ -385,7 +462,18 @@ var PDFDocumentLoadingTask = (function PDFDocumentLoadingTaskClosure() {
      *                   is completed.
      */
     destroy: function () {
-      return this._transport.destroy();
+      this.destroyed = true;
+
+      var transportDestroyed = !this._transport ? Promise.resolve() :
+        this._transport.destroy();
+      var workerDestroyed = transportDestroyed.then(function () {
```
Comment (brendandahl) [inline src/display/api.js]: Uneeded var, just return here.

File: src/display/api.js [hunk 4]
```diff
@@ -994,16 +1082,224 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
...
+          },
+          removeEventListener: function (name, listener) {
+            var i = this._listeners.indexOf(listener);
+            this._listeners.splice(i, 1);
+          },
+          terminate: function WorkerTransport_terminate() {}
+        };
+
+        resolve(fakeWorker);
+      }.bind(this));
```
Comment (brendandahl) [inline src/display/api.js]: Pass in the enclosing reject callback.

File: src/display/api.js [hunk 5]
```diff
@@ -994,16 +1082,224 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
...
+  function setupFakeWorkerGlobal() {
+    if (fakeWorkerReady) {
+      return fakeWorkerReady;
+    }
+
+    fakeWorkerReady = new Promise(function (resolve) {
+      globalScope.PDFJS.disableWorker = true;
+
+      if (!PDFJS.fakeWorkerFilesLoadedCapability) {
+        PDFJS.fakeWorkerFilesLoadedCapability = createPromiseCapability();
```
Comment (brendandahl) [inline src/display/api.js]: I'm confused when this promise ever gets resolved when !PRODUCTION

Comment (brendandahl) [inline src/display/api.js]: Happens outside of the diff, nevermind.

File: src/display/api.js [hunk 6]
```diff
@@ -994,16 +1082,224 @@ var PDFPageProxy = (function PDFPageProxyClosure() {
...
+ * a "fake" worker will be used instead.
+ * @class
+ */
+var PDFWorker = (function PDFWorkerClosure() {
+  var fakeWorkerReady = null;
+  var nextFakeWorkerId = 0;
+
+  // Loads worker code into main thread and setups up the port we will be using
+  // in MessageHandler instances.
+  function setupFakeWorkerGlobal() {
```
Comment (brendandahl) [inline src/display/api.js]: Alot of this is coming from old code, but some of the new parts could be flattened. I don't think there's a reason to create a new promise. Something like https://pastebin.mozilla.org/8852994, in subsequent pull we could make this even better by getting rid of the capability promise.

File: src/display/api.js (hunk 4 above)
Comment (brendandahl) [inline src/display/api.js]: ^not needed if we remove new Promise and flatten.

**Comment by brendandahl** [comment]
r+ if the new promise can be removed and testing run one more time

File: src/display/api.js (hunk 6 above)
Comment (yurydelendik) [inline src/display/api.js]: I actually need to create different instance of the "fakeWorker" for new setupFakeWorkerGlobal request

Comment (yurydelendik) [inline src/display/api.js]: Fixed

**Comment by timvandermeij** [comment]
Nice work!

**Pull request merged**